### PR TITLE
for review of vsphere advanced capabilities

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java [hunk 1]
```diff
@@ -267,6 +267,10 @@ long getClusterId(long vmId) {
             if (destData.getObjectType() == DataObjectType.VOLUME && destStoreTO.getRole() == DataStoreRole.Primary && srcData.getObjectType() == DataObjectType.TEMPLATE
                     && srcStoreTO.getRole() == DataStoreRole.Primary) {
                 needDelegation = false;
+            /* TODO: Uncomment this to disable delegation to SSVM, for copying the template from secondary to primary
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java]: what is the decision to take here @sureshanaparti ? when can we decide? and what do we need for that?

File: plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java [hunk 2]
```diff
@@ -770,6 +804,7 @@ private boolean createVMFullClone(VirtualMachineMO vmTemplate, DatacenterMO dcMo
         return true;
     }
 
+    //TODO: Remove this method after deploying VM from OVF directly using content library
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java]: 👍

Comment (DaanHoogland) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java]: on second thought, we need to be backwards compatible. I don't think we can do this with a single transition script.

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/ContentLibraryHelper.java [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        // Build the storage backing for the library to be created
+        StorageBacking storageBacking = createStorageBacking(context, datastoreName);
+
+        // Build the specification for the library to be created
+        LibraryModel createSpec = new LibraryModel();
+        createSpec.setName(libraryName);
+        createSpec.setDescription("Local content library for datastore " + datastoreName);
+        createSpec.setType(LibraryModel.LibraryType.LOCAL);
+        createSpec.setStorageBackings(Collections.singletonList(storageBacking));
+
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/ContentLibraryHelper.java:96]: sounds like a method should be create called `buildTheSpecificationForTheLibraryToBeCreated(...)`

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/ContentLibraryHelper.java [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+        //Get the storage backing on the datastore
+        StorageBacking dsStorageBacking = createStorageBacking(context, datastoreName);
+        boolean canDelete = false;
+        for (Iterator<StorageBacking> iterator = localLibrary.getStorageBackings().iterator(); iterator.hasNext();) {
+            StorageBacking storageBacking = (StorageBacking) iterator.next();
+            if(dsStorageBacking.equals(storageBacking)) {
+                canDelete = true;
+                break;
+            }
+        }
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/ContentLibraryHelper.java:139]: `getTheStorageBackingOnTheDatastore(...)`

**Review by DaanHoogland — COMMENTED** [review]
thanks Suresh, i'll start a build to deploy and test tomorrow

**Comment by DaanHoogland** [comment]
@sureshanaparti please don't force push on a shared branch. We can clean up before a public PR.

File: plugins/hypervisors/vmware/src/main/java/org/apache/cloudstack/api/command/admin/zone/ImportVsphereStoragePoliciesCmd.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+@APICommand(name = ImportVsphereStoragePoliciesCmd.APINAME, description = "Import vSphere storage policies",
+        responseObject = VsphereStoragePoliciesResponse.class,
+        requestHasSensitiveInfo = false, responseHasSensitiveInfo = false,
+        authorized = {RoleType.Admin})
+public class ImportVsphereStoragePoliciesCmd extends BaseCmd {
+
+    public static final Logger s_logger = Logger.getLogger(ImportVsphereStoragePoliciesCmd.class.getName());
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/main/java/org/apache/cloudstack/api/command/admin/zone/ImportVsphereStoragePoliciesCmd.java]: @harikrishna-patnala , can you make this `static final Logger LOGGER`?

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java [hunk 6]
```diff
@@ -1705,24 +1697,33 @@ public static String removeOVFNetwork(final String ovfString)  {
...
         } catch (SAXException | IOException | ParserConfigurationException | TransformerException e) {
+//            FR37 TODO this warn() should really be an error and the exception should be thrown???
             s_logger.warn("Unexpected exception caught while removing network elements from OVF:", e);
         }
         return ovfString;
     }
 
+    /**
+     * deploys a new VM from a ovf spec. It ignores network, defaults locale to 'US'
+     * @param host
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java]: @DaanHoogland (note to self) remove un-annotated annotation tags

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/PbmPlacementSolverMO.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import com.vmware.vim25.ManagedObjectReference;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PbmPlacementSolverMO extends BaseMO {
+
+    private static final Logger s_logger = Logger.getLogger(PbmPlacementSolverMO.class);
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/PbmPlacementSolverMO.java]: statis final naming

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/PbmProfileManagerMO.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+import org.apache.log4j.Logger;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class PbmProfileManagerMO extends BaseMO {
+
+    private static final Logger s_logger = Logger.getLogger(PbmProfileManagerMO.class);
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/PbmProfileManagerMO.java]: statis final naming

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualStorageObjectManagerMO.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+import com.vmware.vim25.VslmCreateSpecDiskFileBackingSpec;
+import org.apache.log4j.Logger;
+
+import com.vmware.vim25.ManagedObjectReference;
+
+import com.cloud.hypervisor.vmware.util.VmwareContext;
+
+public class VirtualStorageObjectManagerMO extends BaseMO {
+    @SuppressWarnings("unused")
+    private static final Logger s_logger = Logger.getLogger(VirtualStorageObjectManagerMO.class);
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualStorageObjectManagerMO.java]: statis final naming

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/ContentLibraryClient.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+import com.vmware.vapi.core.ExecutionContext.SecurityContext;
+import com.vmware.vapi.protocol.HttpConfiguration;
+import com.vmware.vapi.protocol.ProtocolConnection;
+import com.vmware.vapi.security.SessionSecurityContext;
+
+import com.vmware.vcenter.Datastore;
+import com.vmware.vcenter.ovf.LibraryItem;
+
+public class ContentLibraryClient implements VmwareClientService {
+    private static final Logger s_logger = Logger.getLogger(ContentLibraryClient.class);
```
Comment (DaanHoogland) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/util/ContentLibraryClient.java:46]: statis final naming

**Comment by yadvr** [comment]
@harikrishna-patnala @DaanHoogland can you merge latest master and address the conflicts. I'll start some test/review work in 2 hrs. Thnx.

**Comment by yadvr** [comment]
Nit/tip - merging origin/master is usually faster than doing full rebase ;)

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/manager/ContentLibraryService.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import com.cloud.hypervisor.vmware.util.VmwareContext;
+
+public interface ContentLibraryService {
+    boolean createContentLibrary(VmwareContext context, String primaryDatastoreName) throws Exception;
+
+    boolean deleteContentLibrary(VmwareContext context, String primaryDatastoreName) throws Exception;
+
+    boolean importOvf(VmwareContext context, String sourceOvfTemplateUri, String sourceOvfTemplateName, String targetDatastoreName, String targetOvfTemplateName) throws Exception;
+
+    VirtualMachineMO deployOvf(VmwareContext context, String sourceovfTemplateName, String vmNameToDeploy, VmwareHypervisorHost targetHypervisorHost, DatastoreMO primaryDataStoreMO) throws Exception;
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/manager/ContentLibraryService.java:31]: methods should not throw `Exception` but be more specific

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/manager/ContentLibraryService.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+}
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/manager/ContentLibraryService.java:32]: missing methods:
1. `List<String> listOvfsByTemplateName(VmwareContext context, String partialOvfName)`
2. `boolean deleteOvf(VmwareContext context, String exactOvfName)`

File: plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java [hunk 13]
```diff
@@ -831,6 +831,7 @@ public Answer cloneVolumeFromBaseTemplate(CopyCommand cmd) {
 
             DatastoreMO dsMo = new DatastoreMO(context, morDatastore);
 
+            String cloneVMName = volume.getVmName();
```
Comment (yadvr) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:836]: @nvazquez cc @DaanHoogland - in this method we can actually clone the VM as the instance name is passed to the volume and can be obtained via `volume.getVmName()`... something we can explore/use. If during template register + OVF parsing we know the disks then we can create DB records for all the volumes of the VM (like ROOT-xyz, DATA-xyz-1, DATA-xyz2...) before the StartCommand...

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 14]
```diff
@@ -747,6 +747,10 @@ public void doInTransactionWithoutResult(final TransactionStatus status) {
     public void allocate(final VirtualMachineProfile vm, final LinkedHashMap<? extends Network, List<? extends NicProfile>> networks, final Map<String, Map<Integer, String>> extraDhcpOptions) throws InsufficientCapacityException,
     ConcurrentOperationException {
 
+        if (s_logger.isTraceEnabled()) {
+            // FR37 TODO get getwork requirements from template/vm and compare to networks(.size())
+            s_logger.trace(String.format("allocating networks for %s(template %s); %d networks",vm.getInstanceName(), vm.getTemplate().getUuid(), networks.size()));
+        }
         Transaction.execute(new TransactionCallbackWithExceptionNoReturn<InsufficientCapacityException>() {
             @Override
             public void doInTransactionWithoutResult(final TransactionStatus status) throws InsufficientCapacityException {
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java:756]: I think the use of ` List<NetworkPrerequisiteTO> netprereqs = templateDetailsDao.listNetworkRequirementsByTemplateId(template.getId());` should be on this transaction, and use them to persist nics on DB

File: engine/schema/src/main/resources/META-INF/db/schema-41400to41500.sql [hunk 15]
```diff
@@ -212,10 +212,10 @@ CREATE VIEW `cloud`.`template_view` AS
              OR (`resource_tags`.`resource_type` = 'ISO')))));
 
 -- Add the default offering for deploy-as-is templates. Not visible to the user
-INSERT IGNORE INTO `cloud`.`disk_offering` (name, display_text, customized, unique_name, disk_size, system_use, type, display_offering)
-VALUES ('Custom Deploy-as-is Instance', 'Custom Deploy-as-is Instance', 1, 'Cloud.com-Custom Deploy-as-is Instance', 0, 0, 'Service', 0);
+INSERT IGNORE INTO `cloud`.`disk_offering` (name, uuid, display_text, customized, unique_name, disk_size, system_use, type, display_offering)
+VALUES ('Custom Deploy-as-is Instance', UUID(), 'Custom Deploy-as-is Instance', 1, 'ApacheCloudStack.org-Custom Deploy-as-is Instance', 0, 0, 'Service', 0);
 INSERT IGNORE INTO `cloud`.`service_offering` (`id`, `nw_rate`, `mc_rate`) VALUES
-((SELECT `id` FROM `cloud`.`disk_offering` WHERE `unique_name` = 'Cloud.com-Custom Deploy-as-is Instance'), null, null);
+((SELECT `id` FROM `cloud`.`disk_offering` WHERE `unique_name` = 'ApacheCloudStack.org-Custom Deploy-as-is Instance'), null, null);
```
Comment (DaanHoogland) [inline engine/schema/src/main/resources/META-INF/db/schema-41400to41500.sql]: @shwstppr this change will let you select the custom offering to use as a quick fix.
- UUID as generated in the diskoffering
- uniquename = `ApacheCloudStack.org-Custom Deploy-as-is Instance`

File: engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java [hunk 16]
```diff
@@ -732,7 +734,15 @@ protected void createVolumeFromBaseImageAsync(VolumeInfo volume, DataObject temp
...
         caller.setContext(context);
 
-        motionSrv.copyAsync(context.templateOnStore, volumeOnPrimaryStorage, caller);
+        if (!volume.isDeployAsIs()) {
+            motionSrv.copyAsync(context.templateOnStore, volumeOnPrimaryStorage, caller);
+        } else {
+            Answer answer = new Answer(null);
+            CopyCommandResult result = new CopyCommandResult(null, null);
+            result.setSuccess(true);
+            caller.complete(result);
```
Comment (DaanHoogland) [inline engine/storage/volume/src/main/java/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java:743]: I must say this is a terrible hack, maybe we should rename the method to reflect that the copy may not even be attempted?

**Pull request closed without merging**